querybook/webapp/ui/Comment/Comments.tsx

```diff
@@ -1,184 +1,216 @@
+        currentComment,
+        editComment,
+        editingCommentId,
+        editingCommentParentId,
+    ]);
+    const handleCommentClear = React.useCallback(() => {
+        setEditingCommentId(null);
+        setEditingCommentParentId(null);
+        setCurrentComment(emptyCommentValue);
     }, []);
 
-    const renderThreadCommentDOM = (comment: IComment) => {
-        if (openThreadIds.has(comment.id)) {
-            return comment.child_comments.map(renderFlatCommentDOM);
-        }
-        const uids = Array.from(
-            new Set(comment.child_comments.map((comment) => comment.uid))
-        );
-        return (
-            <>
-                {renderFlatCommentDOM(comment)}
-                <div
-                    className="CommentThread mt12"
-                    onClick={() => handleOpenThread(comment.id)}
+    const renderFlatCommentDOM = (comment: IComment, isChild: boolean) =>
+        comment ? (
+            <Comment
+                key={comment.id}
+                comment={comment}
+                editComment={(text) => handleEditComment(comment.id, text)}
+                isBeingEdited={editingCommentId === comment.id}
+                isChild={isChild}
+                createChildComment={() => setEditingCommentParentId(comment.id)}
+                isBeingRepliedTo={editingCommentParentId === comment.id}
+            />
+        ) : null;
+
+    const handleOpenThread = React.useCallback(
+        (parentCommentId: number, childCommentIds: number[] = []) => {
+            dispatch(
+                fetchChildCommentsByParentCommentIdIfNeeded(
+                    parentCommentId,
+                    childCommentIds
+                )
+            );
+            setOpenThreadIds((curr) => new Set([...curr, parentCommentId]));
+        },
+        [dispatch]
+    );
+
+    const loadingCommentDOM = React.useMemo(
+        () => (
+            <div className="Comment mv8">
+                <StyledText
+                    size="xsmall"
+                    color="lightest"
+                    cursor="default"
+                    isItalic
                 >
+                    Loading Comment
+                </StyledText>
+            </div>
+        ),
+        []
+    );
+
+    const renderThreadCommentDOM = (comment: IComment) => (
+        <>
+            {renderFlatCommentDOM(comment, false)}
+            <div
+                className="CommentThread mt16 mb12"
+                onClick={() =>
+                    handleOpenThread(comment.id, comment.child_comment_ids)
+                }
+            >
+                {openThreadIds.has(comment.id) ? (
+                    comment.child_comment_ids?.map((commentId) =>
+                        commentsById[commentId]
+                            ? renderFlatCommentDOM(
+                                  commentsById[commentId],
+                                  true
+                              )
+                            : loadingCommentDOM
```

Comment:
Is it possible that the api request has been finished, while the `commentsById[commentId]` is still empty, in which case it will keep showing is loading?

Reply:
im not sure i follow

Reply:
what I mean is: what decides when to show `loadingCommentDOM`? 
There might be a state that when the child comments fetching API fails, and check of `commentsById[commentId]` is `false`, it will still show `loadingCommentDOM`, and in which case I think maybe an error should be displayed.